There is only kill

cdata opened the pull request:
Baby's first Rust! This is a crude implementation of killing. Local player can kill everyone. Press `q` to kill.

We probably should have some abstraction for iterating over "other" players, but decided to punt on that.

Fixes #11

cdata committed:
There is only kill

rictic commented on src/lib.rs:
```diff
@@ -148,6 +157,39 @@ impl Game {
+
+        match kill_position {
+            Some(position) => self.kill_player_near(position)?,
+            None => return Ok(()),
```
Since you only want to handle the Some case here, I think you can do:

```rust
if let Some(position) = kill_position {
  self.kill_player_near(position)?;
}
```

rictic commented on src/lib.rs:
```diff
@@ -148,6 +157,39 @@ impl Game {
+            Some(c) => *c,
+        };
+
+        for player in self.players.iter_mut() {
```
Yeah, a way to iterate over other players would be nice.

Writing iterators in Rust today isn't fun, I really want a `yield` construct

rictic commented on src/lib.rs:
```diff
@@ -148,6 +157,39 @@ impl Game {
+                continue;
+            }
+
+            let distance = ((position.0 - player.x).powi(2) + (position.1 - player.y).powi(2))
```
⃤  ⃤  ⃤D ⃤a ⃤m ⃤n ⃤  ⃤y ⃤o ⃤u ⃤  ⃤p ⃤y ⃤t ⃤h ⃤a ⃤g ⃤o ⃤r ⃤a ⃤s ⃤! ⃤  ⃤   ⃤   ⃤

rictic commented on src/lib.rs:
```diff
@@ -148,6 +157,39 @@ impl Game {
+                .sqrt()
+                .abs();
+
+            if distance < self.kill_distance {
```
This kills _a_ nearby player. Would be better to kill the _nearest_ player I think. TODO?

cdata replied:
Yeah, good point. I figured the "rules" of killing are generally TBD, but can do whatever.

mohammadshamma replied:
For the time being it's a battle royal!

rictic commented on src/lib.rs:
```diff
@@ -148,6 +157,39 @@ impl Game {
                 player.x += self.speed * time_steps_passed
             }
+
+            if !inputs.q && self.last_input.q {
```
Am I reading this right and this kills on keyup?

Killing on keydown feels more responsive I think

rictic reviewed (commented):
🔪 !

Nice work!

rictic commented on www/index.js:
```diff
@@ -21,7 +22,14 @@ document.addEventListener('keydown', (ev) => {
     case 'ArrowRight':
       curInput.right = true;
       break;
+    case 'q':
```
just checked, and 'q' is separate from 'Q'

so we should prob handle both here

cdata committed:
Changes based on feedback

 - Kill on key down
 - Handle Q as well as q
 - No need for "last" input
 - Kill closest player first

rictic commented on src/lib.rs:
```diff
@@ -177,6 +175,9 @@ impl Game {
             Some(c) => *c,
         };
 
+        let mut killed_player: Option<&mut Player> = None;
```
ah, nice!

I'm pleasantly surprised that the borrow checker allows this!

cdata replied:
I'm stumbling my way through the dark. I guess I don't know what to expect from the borrow checker at all.

rictic commented on src/lib.rs:
```diff
@@ -177,6 +175,9 @@ impl Game {
         };
 
+        let mut killed_player: Option<&mut Player> = None;
+        let closest_distance = self.kill_distance;
```
Oh, clever way to both get the closest player _and_ restrict based on kill distance!

cdata replied:
:brain:

rictic commented on src/lib.rs:
```diff
@@ -158,14 +157,13 @@ impl Game {
             }
 
-            if !inputs.q && self.last_input.q {
+            if inputs.q {
```
I guess this means that if you hold down kill you just kill everyone nearby.

That's fine though, gotta add a kill cooldown anyways

rictic reviewed (approved):
Nice!
